### ISSUE TITLE
VM: List.removeLast and List.clear leak memory

### ISSUE THREAD

**Issue by ErikCorryGoogle** [dart-lang/sdk#29930]
Using removeLast and clear does not always reduce the memory used by the list.

Should be about 10Mbyte, but uses 14Gbytes.

Here is an example program that uses O(n2) memory where it should only use O(n) memory:

main() {
  List all = [];
  for (int i = 0; i < 50000; i++) {
    List nums = [];
    for (int j = 0; j < i; j++) {
      nums.add(j);
    }
    for (int j = 1; j < i; j++) {
      nums.removeLast();
    }
    if (nums.length == 1) {
      int x = nums[0];
      nums.clear();
      nums.add(x);
    }
    all.add(nums);
  }
}

If the line "nums.clear()" is replaced by "nums = []" then the memory use does not grow to 14Gbytes.

**Comment by lrhn** [comment]
There is a cost to shrinking a list on removes. 
It should definitely not be done on every remove, probably only when the list uses less than 25% of its capacity (which would require removing at least half the elements added since the last time the backing store grew). That still means that every remove gets an extra check (`if (newLength * 4 < _buffer.length) _shrink();`). It will probably still only shrink by half (shrinking exactly to the needed capacity would mean that a single add operation would grow it again). A check on every operation is not a big thing, but it's there.
If you ever fill the `List` to capacity again, you will have to regrow the buffer. Since it's impossible to predict the usage pattern, there will be sequences of operations that become slower if you avoid the memory leak.

One alternative is to only shrink lists on GC. When copying the backing store of a growable list, and the list is <25% used, shrink the list. That won't affect a lot of quick operations on the same list, but if it stays alive for GC, it gets compacted. If you run out of memory, there will be a GC.

For `clear`, it's probably safe to do something more drastic, but again, it can also just be done in preparation for using the list again, otherwise, why keep a reference to the list.

**Comment by ErikCorryGoogle** [comment]
Yes, our initial plan is to use the 25% mark as a good time to shrink.  This should be asymptotically better than what we have now.  We don't expect a performance regression, but it's my opinion that a small constant-factor regression is acceptable to fix an asymptotically worse memory complexity.

Changing the shape and size of objects at GC time is not something I have great experiences with, so I think we want to avoid it.  The problem is that GC happens at a very unpredictable time from the POV of the optimizer, which can easily cause subtle bugs.  Other resizing and rewriting of objects can be tricky (see for example Array::MakeArray), but at least if it happens in a native call or special instruction you have some way to signal to the optimizer that 'interesting stuff happens here'.
